synapse/storage/databases/main/devices.py

```diff
@@ -302,11 +304,17 @@ async def get_device_updates_by_remote(
 
             last_processed_stream_id = update_stream_id
 
+        # len(query_map) + len(cross_signing_keys_by_user) <= len(updates) here,
+        # so len(query_map) + len(cross_signing_keys_by_user) <= limit.
+
         results = await self._get_device_update_edus_by_remote(
             destination, from_stream_id, query_map
         )
 
-        # add the updated cross-signing keys to the results list
+        # len(results) <= len(query_map) here,
+        # so len(results) + len(cross_signing_keys_by_user) <= limit.
```

Comment:
What does "HL-style" mean in the description? Haskell?

Reply:
Hoare Logic, but it's totally irrelevant since I didn't actually go to that much effort to prove it :P

Reply:
time to rewrite synapse in prolog